### PR TITLE
LIMIT pushdown below PROJECT

### PR DESCRIPTION
**Pull request by jeewonhh**
This PR pushes LIMIT below PROJECT for a small constant limit value under 8192.

An example of applicable query is -- `SELECT udf(x) FROM tbl LIMIT 10`

Currently, PROJECT is applied to the entire vector even when the limit value of the subsequent LIMIT operator is smaller than the vector size. The LIMIT pushdown avoids running potentially expensive UDFs or expressions for more tuples than necessary.

LIMIT plays an important role in DuckDB's parallelism and order preservation. LIMIT is now implemented as a sink, which ends the ongoing parallel pipeline and makes the subsequent PROJECT operations execute in a single thread. Therefore, we propose applying this pushdown only for a small limit value - as for a larger limit value, preserving the parallelism will be more helpful than cutting down the number of tuples.

### CODE REVIEW

File: src/optimizer/limit_pushdown.cpp [hunk 1]
```diff
@@ -0,0 +1,42 @@
+#include "duckdb/optimizer/limit_pushdown.hpp"
+
+#include "duckdb/planner/operator/logical_limit.hpp"
+#include "duckdb/planner/operator/logical_projection.hpp"
+#include <iostream>
```
Comment (Tmonster) [inline src/optimizer/limit_pushdown.cpp:5]: nit: don't think we need to include upstream here

File: src/optimizer/limit_pushdown.cpp [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+	}
+	return false;
+}
+
+unique_ptr<LogicalOperator> LimitPushdown::Optimize(unique_ptr<LogicalOperator> op) {
+	if (CanOptimize(*op)) {
+		auto projection = std::move(op->children[0]);
+		op->children[0] = std::move(projection->children[0]);
+		projection->children[0] = std::move(op);
+		swap(projection, op);
```
Comment (Tmonster) [inline src/optimizer/limit_pushdown.cpp:33]: Does the limit get pushed down again? What is you have limit->projection->projection ?

File: src/optimizer/optimizer.cpp [hunk 3]
```diff
@@ -182,6 +183,12 @@ unique_ptr<LogicalOperator> Optimizer::Optimize(unique_ptr<LogicalOperator> plan
 		compressed_materialization.Compress(plan);
 	});
 
+	// pushes LIMIT below PROJECTION
+	RunOptimizer(OptimizerType::LIMIT_PUSHDOWN, [&]() {
```
Comment (Tmonster) [inline src/optimizer/optimizer.cpp:187]: I would put this before compressed materialization. The compressed materialization optimizer introduces extra projection operators. If the limit pushdown is above this, then the optimizer will avoid some extra push downs

**Review by Tmonster — CHANGES_REQUESTED** [review]
Thanks for the PR! Some comments, but otherwise looks nice 👍

File: src/optimizer/limit_pushdown.cpp (hunk 1 above)
Comment (jeewonhh) [reply src/optimizer/limit_pushdown.cpp]: done!

File: src/optimizer/optimizer.cpp (hunk 3 above)
Comment (jeewonhh) [reply src/optimizer/optimizer.cpp:181]: done!

File: src/optimizer/limit_pushdown.cpp (hunk 2 above)
Comment (Tmonster) [reply src/optimizer/limit_pushdown.cpp:32]: Maybe I'm still missing something, but what if the plan looks like
Limit (op) -> Projection -> Projection
After this optimization the plan then looks like

Projection (op) -> Limit -> Projection

But now the optimizer returns. Once if encounters an operator that can be optimized, the children are no longer optimized. 

What I would do is remove the else, and just have 

```
if (CanOptimize(*op)) {
// swap code
}
for (auto &child : op->children) {
	child = Optimize(std::move(child));
}
return op
```

Now if a query is complicated (with many subqueries with limits), this optimization is applied to every limit -> projection pattern.

**Review by Tmonster — CHANGES_REQUESTED** [review]
Can you also add some tests where you inspect the logical plan to ensure that LIMIT is pushed passed the projection? 

`test/optimizer/statistics/statistics_between.test` has some examples where you can REGEX the order of operators in the plan

**Comment by Tmonster** [comment]
Thanks! One more thing, can you also write tests to check that when Limit + offset is used, the result is also correct? 
Thanks!

**Comment by jeewonhh** [comment]
@Tmonster I've added more tests for limit pushdown - could you review this?

**Review by Tmonster — APPROVED** [review]
Thanks! Looks good to me. 
@Mytherin any other thoughts?

**Comment by Tmonster** [comment]
@jeewonhh The PR looks great! DuckDB is ramping up to release 1.0.0 and we are mostly just merging bug fixes. If you want to get this merged to main, you should change the base branch to feature. It will get merged to feature, and once 1.0.0 is released, feature will be merged into main.

**Comment by jeewonhh** [comment]
@Tmonster It's done! Thanks :)

**Comment by jeewonhh** [comment]
@Tmonster Thanks, it's done!

**Comment by jeewonhh** [comment]
@Tmonster Hi, I'm wondering if this will be merged into the feature branch in the near future? Wanted to know the status of my PR :-)

**Comment by Tmonster** [comment]
Hopefully soon! I'll ping mark about it

**Comment by Mytherin** [comment]
Thanks!

**Pull request merged**